File: litedb-python/pd_failover_smoke.py
"""
pd_failover_smoke.py — proves the control plane survives its OWN leader's death.

The placement driver is now a Raft group (`pd.py`). This kills the **PD leader** (which is also a
data node) and asserts:
  1. a NEW PD leader is elected among the surviving PD replicas (the durable decision log carries over);
  2. that new PD leader, with no human action, detects the dead node and re-replicates its shards to
     restore RF — i.e. it finishes the healing the old leader would have done.

So a control-plane leader crash is no longer a SPOF: decisions live in the PD Raft log and any
surviving replica can take over and continue. Run:  python pd_failover_smoke.py
"""

import os

os.environ.setdefault("LITEDB_CLUSTER_NODES", "4")  # 4 nodes: 3 PD + 1 extra, so survivors hold RF 3

import shutil  # noqa: E402
import subprocess  # noqa: E402
import sys  # noqa: E402
import time  # noqa: E402
from collections import Counter  # noqa: E402

import _loader  # noqa: F401,E402
from cluster_client import ClusterClient  # noqa: E402
from cluster_config import DATA_ROOT, INITIAL_NODES, NODES, PD_NODES, SHARDS  # noqa: E402
from rpc import RPCClient  # noqa: E402


def spawn(nid: str):
    return subprocess.Popen([sys.executable, "node.py", nid],
                            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)


def wait_until(fn, timeout=45.0, what=""):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if fn():
            return True
        time.sleep(0.25)
    raise AssertionError("timeout: " + what)


def hosts_of(client):
    out = {}
    for st in client.status():
        if st.get("alive"):
            out[st["node"]] = {sh["group"] for sh in st.get("shards", [])}
    return out


def main():
    n_shards = len(SHARDS)
    shutil.rmtree(DATA_ROOT, ignore_errors=True)
    procs = {n: spawn(n) for n in INITIAL_NODES}
    rpc = RPCClient(timeout=1.5)

    def pd_leader(exclude=()):
        for n in PD_NODES:
            if n in exclude:
                continue
            host, port = NODES[n]
            r = rpc.call(host, port, "pd_status", {})
            res = r.get("result") if r.get("ok") else None
            if isinstance(res, dict) and res.get("ok") and res.get("leader"):
                return res["leader"]
        return None

    try:
        client = ClusterClient()
        wait_until(lambda: pd_leader() is not None, what="initial PD leader elected")
        leader1 = pd_leader()
        print(f"PD Raft group up; leader = {leader1} (PD replicas: {PD_NODES})")

        for i in range(12):
            assert client.put(f"key{i}", f"val{i}").get("ok"), f"put key{i}"
        print("  wrote 12 keys")

        print(f"\nkilling the PD LEADER {leader1} (also a data node) — the control plane must survive")
        procs[leader1].kill()
        procs[leader1].wait()

        wait_until(lambda: pd_leader(exclude={leader1}) not in (None, leader1), timeout=30,
                   what="a new PD leader is elected after the old one died")
        leader2 = pd_leader(exclude={leader1})
        print(f"  new PD leader elected from the surviving replicas: {leader2}")
        assert leader2 and leader2 != leader1

        # the new PD leader must finish the job: detect the dead node and restore RF, no human action
        def healed():
            h = hosts_of(client)
            if leader1 in h:
                return False
            c = Counter(s for ns in h.values() for s in ns)
            return len(c) == n_shards and all(v == 3 for v in c.values())

        wait_until(healed, what="new PD leader re-replicates the dead node's shards to restore RF")
        assert all(client.get(f"key{i}") == f"val{i}" for i in range(12)), "data intact after PD failover"
        print("  RF restored to 3 on survivors; all 12 keys intact")

        print("\nPD FAILOVER OK: the PD leader died; a surviving PD replica took over from the durable "
              "Raft log and finished healing the dead node's data — the control plane is no longer a SPOF.")
    finally:
        for p in procs.values():
            if p.poll() is None:
                p.kill()
        for p in procs.values():
            p.wait()


if __name__ == "__main__":
    main()

File: litedb-java/src/main/java/com/litedb/cluster/PdFailoverSmoke.java
package com.litedb.cluster;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;

/**
 * PdFailoverSmoke — proves the control plane survives its OWN leader's death (mirror of
 * pd_failover_smoke.py). Kill the PD leader (also a data node) and assert (1) a new PD leader is
 * elected among the surviving PD replicas, and (2) it finishes healing the dead node's data to
 * restore RF — with no human action. The control plane is no longer a SPOF.
 *
 * Run with 4 nodes:  LITEDB_CLUSTER_NODES=4 java com.litedb.cluster.PdFailoverSmoke
 */
public class PdFailoverSmoke {

    static Map<String, Process> procs = new LinkedHashMap<>();

    static Process spawn(String nid) throws Exception {
        ProcessBuilder pb = new ProcessBuilder(System.getProperty("java.home") + "/bin/java",
                "-cp", System.getProperty("java.class.path"), "com.litedb.cluster.NodeServer", nid);
        pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);
        pb.redirectError(ProcessBuilder.Redirect.DISCARD);
        Process p = pb.start();
        procs.put(nid, p);
        return p;
    }

    @SuppressWarnings("unchecked")
    static String pdLeader(Rpc.Client rpc, Set<String> exclude) {
        for (String n : ClusterConfig.PD_NODES) {
            if (exclude.contains(n)) continue;
            Map<String, Object> r = rpc.call(ClusterConfig.HOST, ClusterConfig.port(n), "pd_status", NodeServer.resp());
            if (Boolean.TRUE.equals(r.get("ok")) && r.get("result") instanceof Map) {
                Map<String, Object> res = (Map<String, Object>) r.get("result");
                if (Boolean.TRUE.equals(res.get("ok")) && res.get("leader") != null) return (String) res.get("leader");
            }
        }
        return null;
    }

    @SuppressWarnings("unchecked")
    static Map<String, Set<String>> hostsOf(ClusterClient client) {
        Map<String, Set<String>> out = new TreeMap<>();
        for (Map<String, Object> st : client.status()) {
            if (!Boolean.TRUE.equals(st.get("alive"))) continue;
            Set<String> shards = new java.util.HashSet<>();
            for (Object so : (List<Object>) st.get("shards")) shards.add((String) ((Map<String, Object>) so).get("group"));
            out.put((String) st.get("node"), shards);
        }
        return out;
    }

    static boolean allRf3(Map<String, Set<String>> hosts, int nShards) {
        Map<String, Integer> c = new java.util.HashMap<>();
        for (Set<String> s : hosts.values()) for (String sh : s) c.merge(sh, 1, Integer::sum);
        if (c.size() != nShards) return false;
        for (int v : c.values()) if (v != 3) return false;
        return true;
    }

    static void waitUntil(java.util.function.BooleanSupplier c, long ms, String what) throws Exception {
        long deadline = System.currentTimeMillis() + ms;
        while (System.currentTimeMillis() < deadline) { if (c.getAsBoolean()) return; Thread.sleep(250); }
        throw new AssertionError("timeout: " + what);
    }

    public static void main(String[] args) throws Exception {
        int nShards = ClusterConfig.SHARDS.size();
        Path root = Paths.get(ClusterConfig.dataRoot());
        if (Files.exists(root)) Files.walk(root).sorted(Comparator.reverseOrder()).map(Path::toFile).forEach(File::delete);
        Rpc.Client rpc = new Rpc.Client(1500);
        try {
            for (String n : ClusterConfig.INITIAL_NODES) spawn(n);
            ClusterClient client = new ClusterClient();
            waitUntil(() -> pdLeader(rpc, Set.of()) != null, 25000, "initial PD leader elected");
            String leader1 = pdLeader(rpc, Set.of());
            System.out.println("PD Raft group up; leader = " + leader1 + " (PD replicas: " + ClusterConfig.PD_NODES + ")");

            for (int i = 0; i < 12; i++) if (!Boolean.TRUE.equals(client.put("key" + i, "val" + i).get("ok"))) throw new AssertionError("put key" + i);
            System.out.println("  wrote 12 keys");

            System.out.println("\nkilling the PD LEADER " + leader1 + " (also a data node) — the control plane must survive");
            procs.get(leader1).destroyForcibly();
            procs.get(leader1).waitFor();

            final String dead = leader1;
            waitUntil(() -> { String l = pdLeader(rpc, Set.of(dead)); return l != null && !l.equals(dead); },
                    30000, "a new PD leader is elected after the old one died");
            String leader2 = pdLeader(rpc, Set.of(dead));
            System.out.println("  new PD leader elected from the surviving replicas: " + leader2);
            if (leader2 == null || leader2.equals(leader1)) throw new AssertionError("no new PD leader");

            waitUntil(() -> {
                Map<String, Set<String>> h = hostsOf(client);
                return !h.containsKey(dead) && allRf3(h, nShards);
            }, 45000, "new PD leader re-replicates the dead node's shards to restore RF");
            for (int i = 0; i < 12; i++) if (!("val" + i).equals(client.get("key" + i))) throw new AssertionError("data lost after PD failover");
            System.out.println("  RF restored to 3 on survivors; all 12 keys intact");

            System.out.println("\nPD FAILOVER OK (Java): the PD leader died; a surviving PD replica took over from the "
                    + "durable Raft log and finished healing the dead node's data — the control plane is no longer a SPOF.");
        } finally {
            for (Process p : procs.values()) p.destroyForcibly();
            for (Process p : procs.values()) p.waitFor();
        }
        System.exit(0);
    }
}
